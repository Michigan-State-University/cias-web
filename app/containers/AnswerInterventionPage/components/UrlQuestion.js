import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import Question from 'models/Intervention/Question';
import UrlPreview from 'components/UrlPreview';

const UrlQuestion = ({ question, selectAnswer }) => {
  const {
    body: {
      variable: { name },
      data: [{ payload }],
    },
  } = question;

  const changeLinkState = state => {
    selectAnswer([
      {
        var: name,
        payload: state,
      },
    ]);
  };

  useEffect(() => {
    changeLinkState(false);
  }, []);

  const onClick = () => changeLinkState(true);
  return <UrlPreview handleClick={onClick} link={payload} />;
};

UrlQuestion.propTypes = {
  question: PropTypes.shape(Question).isRequired,
  selectAnswer: PropTypes.func,
};

export default UrlQuestion;
