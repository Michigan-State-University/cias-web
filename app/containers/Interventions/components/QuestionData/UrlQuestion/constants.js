export const UPDATE_URL = 'app/EditInterventionPage/URL/UPDATE_URL';
export const UPDATE_VARIABLE = 'app/EditInterventionPage/URL/UPDATE_VARIABLE';
