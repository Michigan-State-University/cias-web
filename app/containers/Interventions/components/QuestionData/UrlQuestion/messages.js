import { defineMessages } from 'react-intl';

export const scope = 'app.containers.UrlQuestion';

export default defineMessages({
  placeholder: {
    id: `${scope}.placeholder`,
    defaultMessage: 'Paste here a web page link to show it to a respondend.',
  },
});
