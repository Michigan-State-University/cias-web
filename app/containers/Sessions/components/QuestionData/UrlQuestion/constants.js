export const UPDATE_URL = 'app/EditSessionPage/URL/UPDATE_URL';
export const UPDATE_VARIABLE = 'app/EditSessionPage/URL/UPDATE_VARIABLE';
