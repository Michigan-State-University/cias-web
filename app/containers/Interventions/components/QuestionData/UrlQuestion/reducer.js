import { UPDATE_URL } from './constants';
import { UPDATE_VARIABLE } from '../constants';

/* eslint-disable default-case, no-param-reassign */
const urlQuestionReducer = (question, payload) => {
  switch (payload.type) {
    case UPDATE_URL:
      const { url } = payload.data;
      question.body.data[0] = { ...question.body.data[0], payload: url };

      return question;

    case UPDATE_VARIABLE:
      const { name } = payload.data;
      question.body.variable.name = name;

      return question;

    default:
      return question;
  }
};

export default urlQuestionReducer;
