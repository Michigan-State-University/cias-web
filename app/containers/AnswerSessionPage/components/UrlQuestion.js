import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';

import { themeColors } from 'theme';

import { QuestionDTO } from 'models/Question';

import UrlPreview from 'components/UrlPreview';
import H3 from 'components/H3';
import Column from 'components/Column';

import messages from '../layouts/messages';

const UrlQuestion = ({ question, selectAnswer }) => {
  const { formatMessage } = useIntl();
  const {
    body: {
      variable: { name },
      data: [{ payload }],
    },
  } = question;

  const changeLinkState = (state) => {
    selectAnswer([
      {
        var: name,
        value: state,
      },
    ]);
  };

  useEffect(() => {
    changeLinkState(false);
  }, []);

  const onClick = () => changeLinkState(true);
  return (
    <Column mt={10}>
      <UrlPreview handleClick={onClick} link={payload} />
      <H3 color={themeColors.warning} textAlign="center">
        {formatMessage(messages.wcagExternalLinkWarning)}
      </H3>
    </Column>
  );
};

UrlQuestion.propTypes = {
  question: PropTypes.shape(QuestionDTO).isRequired,
  selectAnswer: PropTypes.func,
};

export default UrlQuestion;
