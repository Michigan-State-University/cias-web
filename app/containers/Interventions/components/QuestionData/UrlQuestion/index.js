import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { injectIntl } from 'react-intl';

import ApprovableInput from 'components/Input/ApprovableInput';
import Box from 'components/Box';
import Column from 'components/Column';
import Question from 'models/Intervention/Question';
import Row from 'components/Row';
import UrlPreview from 'components/UrlPreview';
import globalMessages from 'global/i18n/globalMessages';
import { BadgeInput } from 'components/Input/BadgeInput';
import { colors } from 'theme/colors';
import { makeSelectDraggable } from 'containers/Interventions/components/QuestionNarrator/selectors';
import { makeSelectSelectedQuestion } from 'containers/Interventions/containers/EditInterventionPage/selectors';
import { updateQuestionData } from 'containers/Interventions/containers/EditInterventionPage/actions';
import { urlValidator } from 'utils/validators/urlValidator';
import { variableNameValidator } from 'utils/validators';

import messages from './messages';
import { UPDATE_URL, UPDATE_VARIABLE } from './constants';

const UrlQuestion = ({
  selectedQuestion,
  updateUrl,
  updateVariable,
  draggable,
  intl: { formatMessage },
}) => {
  const { payload } = selectedQuestion.body.data[0];
  const { variable } = selectedQuestion.body;

  return (
    <Column mt={10}>
      <Row display="flex" hidden={draggable} mb={10}>
        <BadgeInput
          px={0}
          py={12}
          textAlign="center"
          keyboard="tel"
          validator={variableNameValidator}
          placeholder={formatMessage(
            globalMessages.variables.variableNamePlaceholder,
          )}
          value={variable.name}
          color={colors.jungleGreen}
          onBlur={updateVariable}
        />
      </Row>
      <Box
        bg={colors.linkWater}
        width="100%"
        px={21}
        py={14}
        hidden={draggable}
      >
        <Row>
          <ApprovableInput
            rows="3"
            validator={urlValidator}
            placeholder={formatMessage(messages.placeholder)}
            value={payload}
            onCheck={updateUrl}
            type="multiline"
          />
        </Row>
      </Box>
      <Box>{payload && <UrlPreview link={payload} />}</Box>
    </Column>
  );
};

UrlQuestion.propTypes = {
  selectedQuestion: PropTypes.shape(Question).isRequired,
  intl: PropTypes.object.isRequired,
  updateUrl: PropTypes.func.isRequired,
  updateVariable: PropTypes.func.isRequired,
  draggable: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  selectedQuestion: makeSelectSelectedQuestion(),
  draggable: makeSelectDraggable(),
});

const mapDispatchToProps = {
  updateUrl: url => updateQuestionData({ type: UPDATE_URL, data: { url } }),
  updateVariable: name =>
    updateQuestionData({ type: UPDATE_VARIABLE, data: { name } }),
};

export const QuestionUrlWithIntl = injectIntl(UrlQuestion);

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(QuestionUrlWithIntl);
