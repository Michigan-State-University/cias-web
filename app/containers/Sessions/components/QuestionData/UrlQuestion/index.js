import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { injectIntl, FormattedMessage } from 'react-intl';

import { colors, themeColors } from 'theme/colors';
import { urlValidator } from 'utils/validators/urlValidator';
import {
  makeSelectSelectedQuestion,
  updateQuestionData,
} from 'global/reducers/questions';

import ApprovableInput from 'components/Input/ApprovableInput';
import Box from 'components/Box';
import Column from 'components/Column';

import Row from 'components/Row';
import UrlPreview from 'components/UrlPreview';
import Text from 'components/Text';
import { IconTooltip, TooltipType } from 'components/Tooltip';
import Comment from 'components/Text/Comment';

import answerPageMessages from 'containers/AnswerSessionPage/layouts/messages';
import messages from './messages';
import { UPDATE_URL } from './constants';

const UrlQuestion = ({
  selectedQuestion,
  updateUrl,
  isNarratorTab,
  editingPossible,
  intl: { formatMessage },
}) => {
  const { payload } = selectedQuestion.body.data[0];

  const isPreviewValid = urlValidator(payload);
  const displayError = !isPreviewValid && payload;
  const displayPreview = isPreviewValid && payload;

  return (
    <Column mt={10}>
      <Box
        bg={colors.linkWater}
        width="100%"
        px={21}
        py={14}
        hidden={isNarratorTab}
      >
        <Row>
          <ApprovableInput
            disabled={!editingPossible}
            rows="3"
            placeholder={formatMessage(messages.placeholder)}
            value={payload}
            onCheck={updateUrl}
            type="multiline"
          />
        </Row>
      </Box>
      {displayError && (
        <Text mt={15} color={themeColors.warning} fontWeight="bold">
          <FormattedMessage {...messages.invalidUrl} />
        </Text>
      )}
      {displayPreview && (
        <IconTooltip
          id="external-url-warning"
          type={TooltipType.WARNING}
          content={
            <Comment>
              {formatMessage(answerPageMessages.wcagExternalLinkWarning)}
            </Comment>
          }
          stretchContent
        >
          <UrlPreview link={payload} showError />
        </IconTooltip>
      )}
    </Column>
  );
};

UrlQuestion.propTypes = {
  selectedQuestion: PropTypes.object.isRequired,
  intl: PropTypes.object.isRequired,
  updateUrl: PropTypes.func.isRequired,
  isNarratorTab: PropTypes.bool,
  editingPossible: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  selectedQuestion: makeSelectSelectedQuestion(),
});

const mapDispatchToProps = {
  updateUrl: (url) => updateQuestionData({ type: UPDATE_URL, data: { url } }),
};

export const QuestionUrlWithIntl = injectIntl(UrlQuestion);

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(QuestionUrlWithIntl);
