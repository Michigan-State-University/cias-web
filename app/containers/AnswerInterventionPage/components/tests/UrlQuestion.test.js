/**
 *
 * Tests for UrlQuestion
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import { DEFAULT_LOCALE } from 'i18n';

import UrlQuestion from '../UrlQuestion';

describe('<UrlQuestion />', () => {
  const mockedFunctions = {
    selectAnswer: jest.fn(),
  };

  const defaultProps = {
    question: {
      body: {
        variable: { name: '1' },
        data: [{ payload: 'test' }],
      },
      id: 'test-id',
    },
    ...mockedFunctions,
  };

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <MemoryRouter>
          <UrlQuestion {...defaultProps} />
        </MemoryRouter>
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <MemoryRouter>
          <UrlQuestion {...defaultProps} />
        </MemoryRouter>
      </IntlProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
