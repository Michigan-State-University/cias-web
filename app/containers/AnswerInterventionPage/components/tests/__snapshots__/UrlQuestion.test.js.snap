// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UrlQuestion /> Should render and match the snapshot 1`] = `
<div>
  <div
    class="Box-sc-14m3exe-0 hpYCYD"
    width="100%"
  >
    <div
      class="Row-x1o3ru-0 dAhqIQ"
    >
      <div
        class="styled__InlineLoader-tle2yi-1 dIAlHd"
        data-testid="inline-loader"
      >
        <div
          class="styled__StyledSpinner-d123q3-0 iHKSVj"
          color="#C866EA"
          size="50"
          width="6"
        />
      </div>
    </div>
  </div>
</div>
`;
